### PR TITLE
 AC-32928| impl dns group and dns server resiliency coed. tests

### PR DESCRIPTION
**Pull request by yiftahhaimbroad**
implementation of DNS Server resiliency & DNS group resiliency ,  tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1507,3 +1511,70 @@ In addition to arguments above, the following attributes are exported:
 - **last_accessed** (String)
 - **created_on** (String)
 - **modified_on** (String)
+
+# DNS Resiliency resources
+
+Resource: luminate_dns_server_resiliency
+----------
+
+Provides crud of dns resiliency servers
```
Comment (gitomics) [inline README.md:1520]: crud -> CRUD

File: provider/resource_dns_server_resiliency.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		Type:         schema.TypeString,
+		Required:     true,
+		ValidateFunc: utils.ValidateString,
+		Description:  "DNS server name",
+	}
+	DNSServerSchema["group_id"] = &schema.Schema{
+		Type:         schema.TypeString,
+		Required:     true,
+		ValidateFunc: utils.ValidateUuid,
+		Description:  "DNS server name",
```
Comment (gitomics) [inline provider/resource_dns_server_resiliency.go:39]: fix description

File: provider/resource_dns_server_resiliency.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	}
+
+	DNSGroupID := d.Get("group_id").(string)
+	DNSResiliencyServerDTO := &dto.DNSServerInputDTO{
+		Name:            d.Get("name").(string),
+		InternalAddress: d.Get("internal_address").(string),
+		SiteID:          d.Get("site_id").(string),
+		GroupID:         DNSGroupID,
+	}
+	DNSGroup, err := client.DNSResiliencyAPI.GetDNSGroup(DNSGroupID)
```
Comment (gitomics) [inline provider/resource_dns_server_resiliency.go:67]: do we have to get the dns group before calling create?
our BE should be returning an error if there is no such group while we are creating the server, so seems redudndant here in the provider

File: provider/resource_dns_group_resiliency.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		return diag.FromErr(errors.New("invalid client"))
+	}
+	DNSGroupID := d.Id()
+	DNSGroup, err := client.DNSResiliencyAPI.GetDNSGroup(DNSGroupID)
+	if err != nil {
+		log.Println(fmt.Sprintf("[Error] failed Reading DNS Resiliency Group with error: %s", err.Error()))
+		return diag.FromErr(errors.Wrap(err, "read DNS Resiliency group failure"))
+	}
+
+	d.SetId(DNSGroup.ID)
```
Comment (gitomics) [inline provider/resource_dns_group_resiliency.go:84]: how does it make sense we only set the id here?
shouldnt we be copying all the data from BE into the  d *schema.ResourceData

File: provider/resource_dns_server_resiliency.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	DNSResiliencyServerID := d.Id()
+	DNSResiliencyGroupID := d.Get("group_id").(string)
+	_, err := client.DNSResiliencyAPI.GetDNServer(DNSResiliencyServerID, DNSResiliencyGroupID)
+	if err != nil {
+		log.Println(fmt.Sprintf("[Error] failed Reading DNS Resiliency Server with error: %s", err.Error()))
+		return diag.FromErr(errors.Wrap(err, "read DNS Resiliency server failure"))
+	}
+
+	d.SetId(DNSResiliencyServerID)
```
Comment (gitomics) [inline provider/resource_dns_server_resiliency.go:96]: how does it make sense we only set the id here?
shouldnt we be copying all the data from BE into the  d *schema.ResourceData

File: provider/resource_dns_server_resiliency.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+func resourceReadDNSResiliencyServer(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	log.Println("[Info] Reading DNS Resiliency Server")
+	client, ok := m.(*service.LuminateService)
+	if !ok {
+		return diag.FromErr(errors.New("invalid client"))
+	}
+	DNSResiliencyServerID := d.Id()
+	DNSResiliencyGroupID := d.Get("group_id").(string)
+	_, err := client.DNSResiliencyAPI.GetDNServer(DNSResiliencyServerID, DNSResiliencyGroupID)
```
Comment (gitomics) [inline provider/resource_dns_server_resiliency.go:90]: odd that data from BE is not needed in read request..

File: provider/resource_dns_server_resiliency.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+		InternalAddress: d.Get("internal_address").(string),
+		SiteID:          d.Get("site_id").(string),
+	}
+	DNSResiliencyGroupID := d.Get("group_id").(string)
+	_, err := client.DNSResiliencyAPI.UpdateDNServer(DNSResiliencyServerDTO, DNSResiliencyGroupID, d.Id())
+	if err != nil {
+		log.Println(fmt.Sprintf("[Error] failed Updating DNS Resiliency Server with error: %s", err.Error()))
+		return diag.FromErr(err)
+	}
+	d.SetId(d.Id())
```
Comment (gitomics) [inline provider/resource_dns_server_resiliency.go:118]: is SetId needed in each end of operation? shouldnt it be enough in create/delete event?

File: service/dns_resiliency.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+func (d *DNSResiliencyAPI) GetDNSGroup(DNSGroupID string) (*dto.DNSGroupOutputDTO, error) {
+	res, _, err := d.cli.DNSResiliencyApi.GetDNSGroup(context.Background(), DNSGroupID)
+	if err != nil {
+		return nil, err
+	}
+	return dto.ConvertDnsGroupOutputTODTO(res), nil
+}
+
+func (d *DNSResiliencyAPI) ListDNSGroups(input *sdk.DNSResiliencyApiListDnsGroupsOpts) (*sdk.DnsGroupOutputPage, error) {
```
Comment (gitomics) [inline service/dns_resiliency.go:61]: unused, consider removing

File: service/dns_resiliency.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+	res, _, err := d.cli.DNSResiliencyApi.ListDnsGroups(context.Background(), input)
+	if err != nil {
+		return nil, err
+	}
+	return &res, nil
```
Comment (gitomics) [inline service/dns_resiliency.go:66]: not convertign with ConvertDnsGroupOutputTODTO

File: provider/resource_dns_server_resiliency_test.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+resource "luminate_collection_site_link" "new-collection-site-link" {
+	site_id = "${luminate_site.new-site.id}"
+	collection_ids = ["7cef2ccc-ed3e-4812-9ef2-b986c5dac2a5"]
+}
+
+resource "luminate_dns_group_resiliency" "new-dns-group" {
+	name = "testDNSGroupResiliency<RANDOM_PLACEHOLDER>"
+	domain_suffixes = ["%s"]
```
Comment (gitomics) [inline provider/resource_dns_server_resiliency_test.go:26]: odd using %s only for domain suffixes, doesnt look like formatting is really needed here its just aconst string thats being inserted

File: provider/resource_dns_server_resiliency_test.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+		ProviderFactories: newTestAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: strings.ReplaceAll(testDNSServerResiliency, "<RANDOM_PLACEHOLDER>", strconv.Itoa(randNum)),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "name", fmt.Sprintf("testDNSServerResiliency%d", randNum)),
+					resource.TestCheckResourceAttr(resourceName, "internal_address", strings.ReplaceAll("udp://<RANDOM_PLACEHOLDER>.<RANDOM_PLACEHOLDER>.<RANDOM_PLACEHOLDER>.<RANDOM_PLACEHOLDER>:63", "<RANDOM_PLACEHOLDER>", strconv.Itoa(randNum))),
+				),
+				Destroy:     true,
+				ExpectError: regexp.MustCompile(".*"),
```
Comment (gitomics) [inline provider/resource_dns_server_resiliency_test.go:54]: expect any error? sounds odd

if its related to the cant delete last server then maybe we should be expecting just that error

File: provider/resource_dns_server_resiliency_test.go [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+resource "luminate_dns_server_resiliency" "new-dns-server-resiliency" {
+	name = "testDNSServerResiliency<RANDOM_PLACEHOLDER>"
+	site_id = "${luminate_site.new-site.id}"
+	group_id = "${luminate_dns_group_resiliency.new-dns-group.id}"
+	internal_address = "udp://<RANDOM_PLACEHOLDER>.<RANDOM_PLACEHOLDER>.<RANDOM_PLACEHOLDER>.<RANDOM_PLACEHOLDER>:63"
+}
+
+`, "otherdomains<RANDOM_PLACEHOLDER>.com")
+
+func TestAccLuminateDNSServerResiliency(t *testing.T) {
```
Comment (gitomics) [inline provider/resource_dns_server_resiliency_test.go:39]: consider adding a test were only the 2nd server is deleted and not expecting an error

**Review by gitomics — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (yiftahhaimbroad) [reply README.md]: done

File: provider/resource_dns_group_resiliency.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+func resourceCreateDNSResiliencyGroup(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	log.Println("[Info] Creating DNS Resiliency Group")
+	client, ok := m.(*service.LuminateService)
+	if !ok {
+		return diag.FromErr(errors.New("invalid client "))
+	}
+
+	DNSGroupDTO := dto.DNSGroupInputDTO{
+		Name:             d.Get("name").(string),
+		DomainSuffixes:   d.Get("domain_suffixes").([]interface{}),
```
[the remark this answers is not in the document]
Comment (yiftahhaimbroad) [reply provider/resource_dns_group_resiliency.go:59]: tried..

File: provider/resource_dns_group_resiliency.go [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+	if !ok {
+		return diag.FromErr(errors.New("invalid client "))
+	}
+	DNSGroupID := d.Id()
+	DNSGroup, err := client.DNSResiliencyAPI.GetDNSGroup(DNSGroupID)
+	if err != nil {
+		log.Println(fmt.Sprintf("[Error] failed Reading DNS Resiliency Group with error: %s", err.Error()))
+		return diag.FromErr(errors.Wrap(err, "read DNS Resiliency group failure"))
+	}
+	disableDTO := &dto.EnableDisableDNSGroupDTO{
```
[the remark this answers is not in the document]
Comment (yiftahhaimbroad) [reply provider/resource_dns_group_resiliency.go]: removed

File: service/dto/dto.go [hunk 15]
```diff
@@ -260,3 +260,70 @@ func EntityDTOToEntityModel(entities []DirectoryEntity) []sdk.DirectoryEntity {
 	}
 	return directoryEntities
 }
+
+type DNSGroupInputDTO struct {
+	Name             string
+	DomainSuffixes   []interface{} `json:"domainSuffixes"`
```
[the remark this answers is not in the document]
Comment (yiftahhaimbroad) [reply service/dto/dto.go:266]: tried

File: provider/resource_dns_server_resiliency.go (hunk 2 above)
Comment (yiftahhaimbroad) [reply provider/resource_dns_server_resiliency.go]: fixed

File: service/dns_resiliency.go (hunk 8 above)
Comment (yiftahhaimbroad) [reply service/dns_resiliency.go]: removed

File: provider/resource_dns_server_resiliency_test.go (hunk 10 above)
Comment (yiftahhaimbroad) [reply provider/resource_dns_server_resiliency_test.go:26]: agree

File: provider/resource_dns_server_resiliency_test.go (hunk 11 above)
Comment (yiftahhaimbroad) [reply provider/resource_dns_server_resiliency_test.go:54]: working on fix

**Review by yuvalstyrbrod — APPROVED** [review]

**Review by gitomics — APPROVED** [review]

**Pull request merged**